File: solutions/typescript/largest-series-product/1/largest-series-product.ts
export const largestProduct = (digits: string, span: number): number => {
    if (span < 0) {
        throw new Error('Span must not be negative');
    }
    if (span > digits.length) {
        throw new Error('Span must be smaller than string length');
    }
    if (/[^0-9]/.test(digits)) {
        throw new Error('Digits input must only contain digits');
    }

    if (span === 0) {
        return 1;
    }

    let maxProduct = 0;

    for (let i = 0; i <= digits.length - span; i++) {
        const series = digits.substring(i, i + span);
        const product = series
            .split('')
            .map(Number)
            .reduce((acc, n) => acc * n, 1);
        if (product > maxProduct) {
            maxProduct = product;
        }
    }

    return maxProduct;
};

File: solutions/go/largest-series-product/1/largest_series_product.go
package lsproduct

import "errors"

func product(series string) int64 {
	result := int64(1)
	for _, c := range series {
		result *= int64(c - '0')
	}
	return result
}

func LargestSeriesProduct(digits string, span int) (int64, error) {
	if span < 0 {
		return 0, errors.New("span must not be negative")
	}
	digitsLength := len(digits)
	if span > digitsLength {
		return 0, errors.New("span must be smaller than string length")
	}
	for _, c := range digits {
		if c < '0' || c > '9' {
			return 0, errors.New("digits input must only contain digits")
		}
	}

	if span == 0 {
		return 1, nil
	}

	maxProduct := int64(0)

	for i := 0; i <= digitsLength-span; i++ {
		series := digits[i : i+span]
		product := product(series)
		if product > maxProduct {
			maxProduct = product
		}
	}
	return maxProduct, nil
}
